File: a2s1/map/seminar/S2/Seminar2/src/InMemoryRepository.java
import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {
    private Map<ID, E> entities;

    public InMemoryRepository() {
        entities = new HashMap<>();
    }

    @Override
    public E findOne(ID id) {
        if(id == null) {
            throw new IllegalArgumentException("id cannot be null");
        }
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if(entity == null) {
            throw new IllegalArgumentException("entity cannot be null");
        }
        return ;
    }

    @Override
    public E delete(ID id) {
        return null;
    }

    @Override
    public E update(E entity) {
        return null;
    }
}

File: a2s1/map/seminar/S11/ConsoleApp1/Repository/InMemoryRepository.cs
namespace ConsoleApp1.Repository;

public class InMemoryRepository<E,ID> : IRepository<E, ID> where E : Entity<ID>
{
    protected Dictionary<ID,E> Dictionary = new ();
    
    public E? FindOne(ID id)
    {
        return Dictionary.ContainsKey(id) ? Dictionary[id] : default;
    }

    public IEnumerable<E> FindAll()
    {
        return Dictionary.Values;
    }

    public virtual E? Save(E entity)
    {
        return Dictionary.TryAdd(entity.Id, entity) ? entity : null;
    }

    public virtual E? Delete(ID id)
    {
        Dictionary.Remove(id, out E? entity);
        return entity;
    }

    public virtual E? Update(E entity)
    {
        if(!Dictionary.ContainsKey(entity.Id))
            return null;
        Dictionary[entity.Id] = entity;
        return entity;
    }
}